### PR TITLE
matstat: Integer mathematical statistics library

### PR DESCRIPTION
**Pull request by gebart**
### Contribution description

Introduce a new library for estimating parameters of sampled random distributions. The library uses only integer operations and the algorithms only require a single pass over the input data, which reduces the memory usage to O(1).

Includes unit tests of all library functions and for some basic verification of the accuracy of the computed values.

### Issues/PRs references

Required by #8531

### CODE REVIEW

**Comment by gebart** [comment]
I didn't know who to assign, @kYc0o could you reassign to someone you think is suitable for reviewing this PR?

**Comment by kYc0o** [comment]
I can take a look but I don't feel super competent on this matters. Maybe @cladmi ? BTW, we somehow are getting into a policy of not assigning people, but instead assign ourselves (as maintainers) to PRs which we feel we can review. Of course we can still assign someone if we know the person would agree.

**Comment by gebart** [comment]
Yeah, preferably for the review it would be someone with a little experience in numeric algorithms and mathematical statistics, to validate the overall design.

File: sys/include/matstat.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+/**
+ * @brief   Compute the sample variance of all samples so far
+ *
+ * @param[in]   state   State struct to operate on
+ * @param[in]   mean    mean value (e.g. from @ref matstat_mean)
+ *
+ * @return  sample variance
+ */
+uint64_t matstat_variance(const matstat_state_t *state, int32_t mean);
```
Comment (miri64) [inline sys/include/matstat.h]: Why has the mean to be provided? We are able to calculate it from `state` so why not do it in the function?

Comment (gebart) [reply sys/include/matstat.h]: Saves an expensive 64 bit division if you need both mean and variance

**Comment by gebart** [comment]
I found a specific failure case which causes the variance to explode, working on a unit test case for that particular issue

**Comment by gebart** [comment]
Added workaround for a specific truncation error which yielded negative variance in certain situations. A precondition for the fault to occur is that the real variance of the input values is very small, and that a bigger absolute value comes before a smaller value. Also added a test vector for this particular failure mode to catch any future regressions.

File: sys/matstat/matstat.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if (state->count == 0) {
+        state->offset = value;
+    }
+    /* Subtracting this offset yields better numeric stability in the variance
+     * algorithm */
+    value -= state->offset;
+    /* Assuming the differences will be small on average and the number of
+     * samples is reasonably limited, to prevent overflow in sum_sq */
+    state->sum_sq += (int64_t)value * value;
+    ++state->count;
```
Comment (cladmi) [inline sys/matstat/matstat.c]: I would prefer `++state->count` just after the `state->sum += value`.
This way, the first part of the function handles the min/max/mean and the second part the variance and so the `offset` and `sum_sq` values.

File: sys/matstat/matstat.c [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        int32_t mean = matstat_mean(state);
+        /* Adjust offset */
+        matstat_change_offset(state, mean, mean);
+    }
+}
+
+int32_t matstat_mean(const matstat_state_t *state)
+{
+    if (state->count == 0) {
+        /* We don't have any way of returning an error */
```
Comment (cladmi) [inline sys/matstat/matstat.c]: I prefer a note in the documentation than the comment.

File: sys/matstat/matstat.c [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    int32_t mean = state->sum / state->count;
+    DEBUG("mean: %" PRId64 " / %" PRIu32 " = %" PRId32 "\n",
+        state->sum, state->count, mean);
+    return mean;
+}
+
+uint64_t matstat_variance(const matstat_state_t *state, int32_t mean)
+{
+    if (state->count < 2) {
+        /* We don't have any way of returning an error */
```
Comment (cladmi) [inline sys/matstat/matstat.c]: I prefer a note in the documentation than the comment.

File: sys/include/matstat.h [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Internal state for computing running statistics
+ */
+typedef struct {
+    int64_t sum;        /**< Sum of values added */
+    uint64_t sum_sq;    /**< Sum of squared values added */
```
Comment (cladmi) [inline sys/include/matstat.h]: Same as in the functions, I would prefer `sum_sq` defined close to `offset` as they are used together.

File: sys/matstat/matstat.c [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    state->sum += value;
+    if (state->count == 0) {
+        state->offset = value;
+    }
+    /* Subtracting this offset yields better numeric stability in the variance
+     * algorithm */
+    value -= state->offset;
+    /* Assuming the differences will be small on average and the number of
+     * samples is reasonably limited, to prevent overflow in sum_sq */
+    state->sum_sq += (int64_t)value * value;
```
Comment (cladmi) [inline sys/matstat/matstat.c]: Maybe replace the comment by an `assert` and a note in the documentation ?

**Comment by gebart** [comment]
Changing the variance algorithm to a better method, will post back when it is done.

**Comment by gebart** [comment]
Updated the implementation algorithm to Welford's algorithm for variance

File: sys/include/matstat.h (hunk 1 above)
Comment (gebart) [reply sys/include/matstat.h]: A change of algorithm to improve numerical robustness required that the mean be computed for every value added, so I removed the mean function argument

**Comment by kaspar030** [comment]
Looks nice!

(not a bug, thus I'm untagging the release)

**Comment by bergzand** [comment]
@gebart So far the module looks great. I've been playing around a bit and noticed that it is possible to remove the `mean` variable from the `matstat_t` struct by reordering the variance calculations a bit, making use of the fact that the sum and the count are already stored. This does however add a division in the add function and two more divisions in the merge function. Have you noticed the same and do you have any opinion on this? I've opened gebart/RIOT#17 with this code

**Comment by gebart** [comment]
The reason I decided to leave the mean in the struct was to avoid having to perform the same division `sum / count` twice, but maybe it could be worth it to save on RAM. Do you have any benchmarks for the CPU overhead? The 64 bit divisions are quite expensive on Cortex M and other constrained architectures.

**Comment by bergzand** [comment]
Then lets leave it as it is. As you say, 64bit division is not fast, and I think you're right that it's not worth it to reduce the struct size by a `int32_t`.

**Comment by bergzand** [comment]
But no, I can't back this with actual measurements :(

**Comment by gebart** [comment]
there are some corner cases where the computed variance becomes negative, due to limited range of small values. The true variance is close to zero but becomes negative after a number of truncations and roundings make the fractions flip over to the negative side. This can be detected by the sum_sq variable suddenly becoming very big.
I am considering changing the variance and sum_sq variables to signed int and letting it show negative variance when this happens, it can simply be discarded as "close to zero" when interpreting the results. It kind of messes up the printouts in many cases when the variance becomes extremely large when using unsigned ints.
A different approach is to check for underflow and truncate the sum to zero when it becomes negative.

**Comment by gebart** [comment]
@bergzand sorry, I was maybe a bit too quick when I replied before. The mean is only a 32 bit division, maybe it is worth the RAM savings to recompute on every add? I don't know. Perhaps we should revisit this after merging the basic implementation?

**Comment by bergzand** [comment]
> there are some corner cases where the computed variance becomes negative, due to limited range of small values. The true variance is close to zero but becomes negative after a number of truncations and roundings make the fractions flip over to the negative side. This can be detected by the sum_sq variable suddenly becoming very big.

Can you provide a test case for this?

Is it possible to wrap the `sum_sq` addition in an absolute operator:
`state->sum_sq = abs (sum_sq + (value - state->mean) * (value - new_mean));`, or does this screw too much with the results?

**Comment by bergzand** [comment]
> A different approach is to check for underflow and truncate the sum to zero when it becomes negative.

Oh, my suggestion is essentially something like this :)

**Comment by gebart** [comment]
OK to squash?

**Comment by bergzand** [comment]
> OK to squash?

:+1:

**Comment by gebart** [comment]
For future reference, here is a test input for the negative variance tests.
Start with an empty matstat_state_t and use matstat_merge to successively merge each of the given states in order and you should end up with a negative variance like I did below.
```
2018-03-22 10:44:07,610 - INFO #    interval    count       sum       sum_sq    min   max  mean  variance
2018-03-22 10:44:07,612 - INFO #   16 -   17:    2686      5414         1380      1     3     2      0
2018-03-22 10:44:07,615 - INFO #   18 -   19:    2643      5272         3263      1     3     1      1
2018-03-22 10:44:07,627 - INFO #   20 -   23:    2650      5328          719      1     3     2      0
2018-03-22 10:44:07,629 - INFO #   24 -   31:    2562      5117         2756      1     3     1      1
2018-03-22 10:44:07,641 - INFO #   32 -   47:    2579      5157          635      1     3     1      0
2018-03-22 10:44:07,644 - INFO #   48 -   79:    2533      5050         2944      1     3     1      1
2018-03-22 10:44:07,646 - INFO #   80 -  143:    2630      5276         1078      1     3     2      0
2018-03-22 10:44:07,658 - INFO #  144 -  271:    2667      5333          974      1     3     1      0
2018-03-22 10:44:07,661 - INFO #  272 -  527:    2414      4859         1074      1     3     2      0
2018-03-22 10:44:07,673 - INFO #       TOTAL    23364     46806       -11106      1     3     2 789570863061659  <=== SIC!

2018-03-22 10:58:57,213 - INFO #    interval    count       sum       sum_sq    min   max  mean  variance
2018-03-22 10:58:57,215 - INFO #   16 -   17:   55988    294157        63150      3     7     5      1
2018-03-22 10:58:57,227 - INFO #   18 -   19:   55785    274856        99726      3     6     4      1
2018-03-22 10:58:57,230 - INFO #   20 -   23:   55833    237601        25047      3     6     4      0
2018-03-22 10:58:57,233 - INFO #   24 -   31:   55919    216642        48885      3     4     3      0
2018-03-22 10:58:57,244 - INFO #   32 -   47:   56151    217615        49162      3     4     3      0
2018-03-22 10:58:57,248 - INFO #   48 -   79:   55743    216097        48857      3     4     3      0
2018-03-22 10:58:57,251 - INFO #   80 -  143:   56212    217884        49247      3     4     3      0
2018-03-22 10:58:57,261 - INFO #  144 -  271:   56090    217362        49088      3     4     3      0
2018-03-22 10:58:57,265 - INFO #  272 -  527:   52853    204947        46381      3     4     3      0
2018-03-22 10:58:57,276 - INFO #       TOTAL   500574   2097161      -516847      3     7     4 36851256607346  <=== SIC!

```

**Comment by bergzand** [comment]
@gebart To get this merged, maybe add a small warning in the header about rounding because of integers and to expect weird behaviour when using it with many small valued numbers. This can be added to the headers file [here](https://github.com/RIOT-OS/RIOT/pull/8733/files#diff-1c5158497f5c6e373364f5e9862fa439R25)

Is this okay with you or do you want to hold this until you have a fix for this corned case?

**Comment by gebart** [comment]
and add that log file data as a unit test

**Comment by gebart** [comment]
- Added a regression test for negative variance after merge
- Add safeguard against negative variance to fix the failing test.
- Update comment on sigma2 as suggested by @cladmi 
- Rebased

**Comment by kaspar030** [comment]
Seems like this just needs squashing.

**Comment by gebart** [comment]
Squashed

**Comment by gebart** [comment]
@cladmi @kaspar030 @bergzand @miri64 @kYc0o Thanks for the help getting this merged!

**Pull request merged**